x-pack/plugins/security_solution/public/onboarding/components/onboarding_body/cards/common/card_link_button.tsx

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useCallback } from 'react';
+import type { OnboardingCardId } from '../../../../constants';
+import { SecuritySolutionLinkButton } from '../../../../../common/components/links';
+import { useOnboardingContext } from '../../../onboarding_context';
+
+export interface CardLinkButtonProps {
+  linkId: string;
+  cardId: OnboardingCardId;
+}
+
+export const withReportCardLinkClick = <T extends { onClick?: React.MouseEventHandler }>(
```

Comment:
we are in a file for CardLinkButton, yet we export a HoC-ed version of another component, and this doesn't seem to be a shared generic HoC either, so to me this indicates that we should turn this HoC into a component, which should make this 2x more readable and we hopefully won't need to marry carried types. wdyt?

Reply:
Not sure I understand. This file exports the `CardLinkButton` component, which is a composed ("Security Solution" generic) button component, which is at the same time composed of EuiButton. Here we are adding onboarding-page-specific props to integrate "click telemetry" functionality.

It also exports the `withReportCardLinkClick` HOC in case we need to use any other kind of link component (anchor, IconButton...) and we want to have integrated "click telemetry". The only prop the wrapped component needs to have is the `onClick` callback.

Could you clarify the suggestion?

Reply:
Sure. So what I was saying is you have a card_link_button file, and I usually expect to see a component there. But instead I see a HoC. HoCs are usually additional part next to the exported component.

Also you are making an attempt to marry these types which is hard to do in a HoC usually also it renders it increasingly hardly readable with all those returning functions and destructures etc.

So I was thinking why not just turn the HoC into a component like:

```tsx
export interface CardLinkButtonProps
  extends React.ComponentProps<typeof SecuritySolutionLinkButton> {
  linkId: string;
  cardId: OnboardingCardId;
}

export const CardLinkButton: React.FC<CardLinkButtonProps> = React.memo(
  ({ onClick, cardId, linkId, ...rest }) => {
    const { reportCardLinkClicked } = useOnboardingContext();
    const onClickWithReport = useCallback<React.MouseEventHandler>(
      (ev) => {
        reportCardLinkClicked(cardId, linkId);
        onClick?.(ev);
      },
      [reportCardLinkClicked, cardId, linkId, onClick]
    );

    return <SecuritySolutionLinkButton onClick={onClickWithReport} {...rest} />;
  }
);

CardLinkButton.displayName = 'CardLinkButton';
```

So that's that.

Now you say maybe this is going to be reused in future. Which is fine. And what we can do here we can actually extract the meat of the render into a shared hook and call in whatever other component will need this functionality that shares only onClick.

What I am trying to say is that HoCs were a good solution pre typescript and pre hooks for class based architecture. But recently especially for complex cases they seemingly add more complexity typewise and make it less readable, imho, especially given that hooks can achieve just the same. So I try to abstain from them as much as possible where possible.

And to just save you some time here is the hook the approach you could use:

```tsx
// shared/useCardLinkReportClick.tsx
export interface UseCardLinkReportProps {
  onClick?: React.MouseEventHandler;
  linkId: string;
  cardId: OnboardingCardId;
}

export const useCardLinkReportClick = ({ cardId, linkId, onClick }: UseCardLinkReportProps) => {
  const { reportCardLinkClicked } = useOnboardingContext();
  return useCallback<React.MouseEventHandler>(
    (ev) => {
      reportCardLinkClicked(cardId, linkId);
      onClick?.(ev);
    },
    [reportCardLinkClicked, cardId, linkId, onClick]
  );
};

// CardLinkButton.tsx
export interface CardLinkButtonProps
  extends React.ComponentProps<typeof SecuritySolutionLinkButton> {
  linkId: string;
  cardId: OnboardingCardId;
}

export const CardLinkButton: React.FC<CardLinkButtonProps> = React.memo(
  ({ onClick, cardId, linkId, ...rest }) => {
    const onClickWithReport = useCardLinkReportClick({ cardId, linkId, onClick });

    return <SecuritySolutionLinkButton onClick={onClickWithReport} {...rest} />;
  }
);

CardLinkButton.displayName = 'CardLinkButton';
```

This way you imho keep it all straightforward and readable. Im sure if you come back to reading this a couple of months later you will thank yourself. :)

let me know what you think and whether I missed anything

Reply:
Yes, but if you need this same functionality in another component, let's say a `SecuritySolutionLinkAnchor`, you need to duplicate the exact same code in a separate component and use a different component at the end. 
With the HOC:

`export const CardLinkAnchor = withReportCardLinkClick(SecuritySolutionLinkAnchor);`

The `CardLinkButton` component is exported from this file, which is all the consumer cares about. I don't quite understand the concern with types, it is the same implementation but encapsulated in a generic function.

I can agree that it would be better to put this HOC in a separate `with_report_card_link_click.ts` file. But we can do that later if we need to use it with more components.

Reply:
Also you have a type security hole here
![image](https://github.com/user-attachments/assets/a3837bf6-ea29-498a-aff2-7d8b255b42cc)
which is more and more inevitable with HoCs and type casting.

To me readability and type security trumps DRY. But I see the value in your point as well.

And this is a nit: forgot to add. So feel free to resolve.

Reply:
> Also you have a type security hole here

Sure, the `as unknown as T` is necessary because of TS limitations, you can hardcode whatever you want there. But the returned component props are typed, consumers can not pass arbitrary props there.

Reply:
> Sure, the as unknown as T is necessary because of TS limitations, you can hardcode whatever you want there. But the returned component props are typed, consumers can not pass arbitrary props there.

Yes and that's the danger. Because if a consumer messes up their code it's localized. But if we or next person accidentally mess up something here later and there is no type check to catch it. It will blow up for all consumers at once and consumer code will not catch it because from the outside they trust the shape to be what they are told.

Reply:
> Sure, the as unknown as T is necessary because of TS limitations, you can hardcode whatever you want there. But the returned component props are typed, consumers can not pass arbitrary props there.

Yes and that's the danger. Because hypothetically if a consumer would mess up their code it'd be localized. But in this case as you said consumer props are properly typed, but we can still break it upon reception from here thanks to HoC and  type safety hole. If we or next person maintaining this accidentally mess up something here later and there is no type check to catch it, it might override consumer's properly typed code and It will blow up for all consumers at once and consumer code will not catch it because from the outside they fulfilled the perceived contract. And that's the conniving part here. That's why type casts should be avoided. And I believe this is not a limitation and we can solve this but it will consume mental cycles and will potential make the code uglier.